### PR TITLE
feat: gather v0 block signatures from stackerdb

### LINKED ISSUE #4775: Nakamoto Signer[3.0] -  The sign_coordinator must be updated to send a block proposal message rather than initiate a signing round
stacks-node: nakamoto_node::sign_coordinator

The sign_coordinator must be updated to send a block proposal message rather than initiate a signing round (with a NonceRequest). It will listen on the stackerdb for signer responses, as it does currently, but the signer responses will be `MessageSignature`s, rather than WSTS packets.

The sign_coordinator will need to validate the signatures and tally the results. Once the validation threshold is reached, they can announce the block.

### PR DESCRIPTION
**Pull request by hstove**
- Closes #4775 

This PR updates the miner's `SignCoordinator` to gather signer signatures using the new header format.

### CODE REVIEW

**Comment by hstove** [comment]
Opening for draft reviews - there is one integration test implemented that demonstrates the full flow of proposing a Nakamoto block, gathering block responses from signers, and then broadcasting that block.

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (hstove) [inline testnet/stacks-node/src/nakamoto_node/miner.rs:177]: I moved the loading of the reward set out of `gather_signatures` and into this higher-level spot so that we're not loading the reward set on every single loop. I see that the code above this is mainly using `expect`s instead of returning errors, so I followed that same pattern here, but wanted to double check that that's right

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/miner.rs:163]: Let's not make any of these calls fatal.  The miner code is constantly re-trying to build blocks, so anything the miner does should be treated as fallible.

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/miner.rs:177]: This `.expect()` is reachable in the node's normal mode of operation, since the node may first store a `SelectedAndUnknown` reward set which will get updated to `SelectedAndKnown` once the anchor block is processed.

Comment (jcnelson) [reply testnet/stacks-node/src/nakamoto_node/miner.rs:177]: Let's remove these `.expect`s then.  The miner is going to be retrying this code over and over, so any transient problems (like my comment below) might go away later.

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/miner.rs:329]: Aren't we removing the aggregate public key?

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:203]: Let's just remove the aggregate public key.  It's not going to be used for some time.

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:317]: Let's just go ahead and strip this code out for now.  Or, if there's test code that uses it, then let's put it into the test module directly.

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:721]: This should just return an error.

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:711]: I don't think we want to do this here.  If we're mutating the bitvec _after_ we get a signer signature, then it would mean that the signer would need to re-sign the block with the new bitvec.

Instead, the miner needs to obtain the signer-determined bitvec _a priori_, and just put it directly into the block.  How exactly this happens is still undetermined (it'll be a separate PR that can ship after Nakamoto), but the miner shouldn't be in the business of modifying the signer bitvec due to the above.

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:203]: Ok, can do. For things like `begin_sign_v1`, should we just use `Point::new()` with a TODO?

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:317]: There is test code that could use it (ie `tests/signer/v1`), but until we have an epoch-gated approach, the miner won't ever trigger this code.

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:711]: Yeah, I had added this in the `begin_sign_v1` code a while back, and the purpose was to set up the ability for the miner to potentially use it in the _next_ Nakamoto block. If I'm thinking the same way, I think that would be the "a priori" approach, but only utilized in the next block. This `next_signers_bitvec` is never actually used, though, so I'm fine with taking it out.

There are lots of gotchas here - like the signer's signature could come in right after the miner reaches the threshold, and so it doesn't really make sense to remove them from the Bitvec in the next block, because that signer is just going to reject the block. This is actually a good potential use case for `next_signer_bitvec` - the miner got a response from the signer, but the block was rejected, so the miner retries the same block with an update BitVec.

Overall fine with punting on this, as this is just an unused feature at the moment.

Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:711]: Context from the PR that added this: https://github.com/stacks-network/stacks-core/pull/4627

**Comment by jferrant** [comment]
We do have a [ticket](https://github.com/stacks-network/stacks-core/issues/4802) in place to port some existing signer tests over that might be relevant but in terms of testing the miner signature aggregation, this seems like a good test and I wouldn't block this PR based on these other tests. 

I do think we should add some unit tests to the logic for accumulating signatures. It might be a bit difficult as you may need to forcibly write stackerdb events, but we should simulate bad signaurtes being written and check that threshold calculations are as expected.

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/miner.rs]: Removed and moved into `gather_signatures`, which will gracefully re-loop if an error occurs

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs]: Removed 👍🏼

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs]: Removed 👍🏼

**Comment by hstove** [comment]
@jferrant yeah good call, I'll add more logic and tests around verifying and gathering signatures

**Comment by hstove** [comment]
@jcnelson @kantai I've opened this as "ready for review" after adding additional logic and checks for the data the miner gets from StackerDB. I've really struggled with integration testing for these scenarios. For example, when using `SignerTest`, when stopping a signer it stops the event observer, which hangs the whole process.

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (kantai) [inline testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:265]: I think this would be nice to implement as an instance method in `RewardSet`, i.e.:

```rust
pub fn total_signing_weight(&self) -> Result<u32, ChainstateError> {
       let Some(ref reward_set_signers) = self.signers else {
            error!("Could not initialize WSTS coordinator for reward set without signer");
            return Err(ChainstateError::NoRegisteredSigners(0));
       };
       reward_set_signers
                .iter()
                .fold(0, |s, acc| {
                    acc.checked_add(s.weight)
                        .expect("FATAL: Total signer weight > u32::MAX")
                })
}
```

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (kantai) [inline testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:763]: This might be helpful with a bit more info:

```rust
debug!("Added another signature to block";
             "block_signer_sighash" => %block_sighash,
             "signer_pubkey" => %signer_pubkey,
             "signer_slot_id" => slot_id,
             "signature" => %signature,
             "signer_weight" => signer_entry.weight 
             "total_weight_signed" => total_weight_signed,
             );
```

**Review by kantai — COMMENTED** [review]
This LGTM, just a few comments

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs]: Good call - I've added this, but I removed `signer_weight` as it otherwise goes over the "max recursion" limit.

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs]: 👍🏼 , added and am now using it in `SignCoordinator::new` and `verify_signer_signature`

**Comment by kantai** [comment]
Oh, that’s just because the comma was missing. Can you add it back in?

**Comment by hstove** [comment]
Ah, yeah added 😀

**Review by kantai — APPROVED** [review]
LGTM!

**Review by jferrant — APPROVED** [review]

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:278]: Can this be moved into the `#[cfg(test)]` block?  Also, can the imports for wsts be moved there as well?

File: testnet/stacks-node/src/tests/signer/v0.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/signer/v0.rs:300]: It this getting tested in CI?

**Review by jcnelson — APPROVED** [review]
LGTM; just a couple of minor comments

File: testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs
Comment (hstove) [reply testnet/stacks-node/src/nakamoto_node/sign_coordinator.rs:278]: Both the `cfg(test)` block and the "non test" block return a `SignCoordinator`, which has a `coordinator` field. So, we'd have to return some other type if we wanted to only include the wsts stuff in the `cfg(test)` block. Ultimately, this whole file relies on both the v0 and v1 stuff (ie `begin_sign_v0` and `begin_sign_v1`) - there's no easy fix other than separate files / structs IMO

File: testnet/stacks-node/src/tests/signer/v0.rs
Comment (hstove) [reply testnet/stacks-node/src/tests/signer/v0.rs:300]: This test is running in the `integration-tests` workflow, and I believe those tests do include the `monitoring_prom` flag enabled.

**Comment by hstove** [comment]
I'm going to merge this - @jcnelson if you want us to address your comment regarding the `cfg(test)` section, feel free to bring it back up in #4781

**Pull request merged**
